common/lib/xmodule/xmodule/tabs.py

```diff
@@ -168,610 +161,30 @@ def from_json(tab_dict):
-
-
-class EdxNotesTab(AuthenticatedCourseTab):
-    """
-    A tab for the course student notes.
-    """
-    type = 'edxnotes'
-
-    def can_display(self, course, settings, is_user_authenticated, is_user_staff, is_user_enrolled):
-        return settings.FEATURES.get('ENABLE_EDXNOTES')
-
-    def __init__(self, tab_dict=None):
-        super(EdxNotesTab, self).__init__(
-            name=tab_dict['name'] if tab_dict else _('Notes'),
-            tab_id=self.type,
-            link_func=link_reverse_func(self.type),
-        )
-
-    @classmethod
-    def validate(cls, tab_dict, raise_error=True):
-        return super(EdxNotesTab, cls).validate(tab_dict, raise_error) and need_name(tab_dict, raise_error)
-
-
-class InstructorTab(StaffTab):
-    """
-    A tab for the course instructors.
-    """
-    type = 'instructor'
-
-    def __init__(self, tab_dict=None):  # pylint: disable=unused-argument
-        super(InstructorTab, self).__init__(
-            # Translators: 'Instructor' appears on the tab that leads to the instructor dashboard, which is
-            # a portal where an instructor can get data and perform various actions on their course
-            name=_('Instructor'),
-            tab_id=self.type,
-            link_func=link_reverse_func('instructor_dashboard'),
-        )
-
-
-class CcxCoachTab(CourseTab):
-    """
-    A tab for the custom course coaches.
-    """
-    type = 'ccx_coach'
-
-    def __init__(self, tab_dict=None):  # pylint: disable=unused-argument
-        super(CcxCoachTab, self).__init__(
-            name=_('CCX Coach'),
-            tab_id=self.type,
-            link_func=link_reverse_func('ccx_coach_dashboard'),
-        )
-
-    def can_display(self, course, settings, *args, **kw):
-        """
-        Since we don't get the user here, we use a thread local defined in the ccx
-        overrides to get it, then use the course to get the coach role and find out if
-        the user is one.
-        """
-        user_is_coach = False
-        if settings.FEATURES.get('CUSTOM_COURSES_EDX', False):
-            from opaque_keys.edx.locations import SlashSeparatedCourseKey
-            from student.roles import CourseCcxCoachRole  # pylint: disable=import-error
-            from ccx.overrides import get_current_request  # pylint: disable=import-error
-            course_id = course.id.to_deprecated_string()
-            course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
-            role = CourseCcxCoachRole(course_key)
-            request = get_current_request()
-            if request is not None:
-                user_is_coach = role.has_user(request.user)
-        super_can_display = super(CcxCoachTab, self).can_display(
-            course, settings, *args, **kw
-        )
-        return user_is_coach and super_can_display
+            return None
+        tab_type.validate(tab_dict)
+        return tab_type.create_tab(tab_dict=tab_dict)
```

Comment:
After talking with @andy-armstrong , we both think that it makes sense to leave them as separate steps. It's a little weird, but since this is one of the single points of tab generation, it's not going to spill out into other things too much.